feat: add AutoEvents docs (#1333)
edgexfoundry/edgex-docs: docs_src/microservices/device/details/AutoEvents.md

---

# Device Service - Auto Events
## Concept
![image](Auto_Events.png)

> In the Operational Technology (OT) working field, there's a need to check/menitor the status of the actual device/sensor, for example, when the temperature reaches 30°C, the Air Conditioner will turn on itself. However, it is time-consuming to check sensor's status manually so edgex provides this functionality called <code>AutoEvents</code> to check their statuses automatically and periodically.

Reviewer: ```suggestion
In the Operational Technology (OT) working field, there's a need to check/monitor the status of the actual device/sensor, for example, when the temperature reaches 30°C, the Air Conditioner will turn on itself. However, it is time-consuming to check sensor's status manually so edgex provides this functionality called <code>AutoEvents</code> to check their statuses automatically and periodically.
```

Author: https://iotechsys.atlassian.net/wiki/spaces/CP/pages/3611197868/Jinko+DCB-3440+Modbus+Map

Author: updated

> From the diagram above, <code>Autoevents</code> is used to define how often events/readings are collected to be sent to core data from the device service. This is an optional feature when creating a deviceso each device may or may not have multiple autoevents associated with it. An AutoEvent has the following fields:

Reviewer: The diagram doesn't work for me. Why is Core Metadata sending reading to Core Data.

Device service should "Get" reading(s) from device and send event/reading to MessageBus when App service/Core Data will receive them.

Author: thanks Lenny, updated

> From the diagram above, <code>Autoevents</code> is used to define how often events/readings are collected to be sent to core data from the device service. This is an optional feature when creating a deviceso each device may or may not have multiple autoevents associated with it. An AutoEvent has the following fields:

Reviewer: ```suggestion
From the diagram above, <code>Autoevents</code> is used to define how often events/readings are collected to be sent to core data from the device service. This is an optional feature when creating a devices each device may or may not have multiple autoevents associated with it. An AutoEvent has the following fields:
```

Author: updated

> - **resource**: the name of a deviceResource or deviceCommand indicating what to read.

Reviewer: ```suggestion
- **source**: the name of a deviceResource or deviceCommand indicating what to read.
```

Author: updated

> - **frequency**: a string indicating the time to wait between reading events, expressed as an integer followed by units of ms, s, m or h.

Reviewer: ```suggestion
- **frequency**: a string indicating the time to wait between readings, expressed as an integer followed by units of ms, s, m or h.
```

Author: updated

- **onchange**: a boolean: if set to true, only generate new events if one or more of the contained readings has changed since the last event.

## Example and Usage

The AutoEvent is defined in the `autoEvents` section of the device definition file:

> After service startup, query core-data's API. The results show that the service auto-executes the command every 30 seconds.

Reviewer: This is confusing. You state to query  core-data's API and then show a partial device definition file. I would move the example YAML (complete device definition) above this  paragraph and use `!!! example` 

Then be more specific on which core-data api to use to display the event/reads generated with example data.

From the diagram above, <code>Autoevents</code> is used to define how often events/readings are collected to be sent to core data from the device service. When the device service receives the information of devices being created, updated or deleted from core metadata, it will get reading from the real device automatically and send event/reading to MessageBus when App service/Core Data receive them.

<code>Autoevents</code> is an optional feature when creating a devices, each device may or may not have multiple autoevents associated with it. An AutoEvent has the following fields:

- **sourceName**: the name of a deviceResource or deviceCommand indicating what to read.
- **interval**: a string indicating the time to wait between reading, expressed as an integer followed by units of ms, s, m or h.

> - **onchange**: a boolean: if set to true, only generate new events if one or more of the contained readings has changed since the last event.

Author: updated

## Example and Usage

The AutoEvent is defined in the `autoEvents` section of the device definition file:

!!! example
    ```yaml
    device:
      name: device-demo
      adminState: UNLOCKED
      operatingState: UP
      serviceName: device-virtual
      profileName: virtual-profile
      protocols: virtual
      autoEvents:
      - interval: 30s 
        onChange: false
        sourceName: temperature
    ```

After service startup, query core-data's API to query the number of events:

``` bash

> curl http://localhost:59880/api/v3/event/all

Reviewer 2: use the API "/event/device/name/{name}" instead of all

Author: updated

```

The results show that the service auto-executes the command every 30 seconds:

> 

Reviewer 2: use a clear example to show the auto event result.  one event doesn't show the service auto-executes the command every 30 seconds.

Author: updated

# Device Service - Auto Events
## Concept

> ![image](Auto_Events.png)

Reviewer: In the new diagram, the second Core Metadata on the right should be App Services.

Author: oh yeah absolutely, thanks! updated
